### PR TITLE
[3] Support "endpoints" in deployments server YAML conf

### PR DESCRIPTION
**Pull request by dbczumar**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/dbczumar/mlflow/pull/10467?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10467/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10467
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Support "endpoints" in deployments server YAML conf

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [X] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/gateway/config.py [hunk 1]
```diff
@@ -313,10 +314,23 @@ def validate_config(cls, config, values):
...
+    """
+    Enables use of field aliases in a configuration model for backwards compatibility
+    """
+
+    if Version(pydantic.__version__) >= Version("2.0"):
+        model_config = ConfigDict(populate_by_name=True)
+    else:
+
+        class Config:
+            allow_population_by_field_name = True
```
Comment (dbczumar) [inline mlflow/gateway/config.py:327]: Otherwise pydantic prints an annoying warning

File: mlflow/deployments/cli.py [hunk 2]
```diff
@@ -490,6 +490,6 @@ def validate_config_path(_ctx, _param, value):
     help="The number of workers.",
 )
 def start_server(config_path: str, host: str, port: str, workers: int):
-    from mlflow.gateway.runner import run_app
```
Comment (dbczumar) [inline mlflow/deployments/cli.py:493]: This was broken...I'll add some test coverage

File: mlflow/gateway/utils.py [hunk 3]
```diff
@@ -30,13 +30,14 @@ def is_valid_endpoint_name(name: str) -> bool:
 
 
 def check_configuration_route_name_collisions(config):
-    if len(config["routes"]) < 2:
+    routes = config.get("routes", []) or config.get("endpoints", [])
```
Comment (harupy) [inline mlflow/gateway/utils.py:33]:
```suggestion
    routes = config.get("routes", config.get("endpoints", []))
```

File: examples/gateway/openai/config.yaml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-routes:
+endpoints:
   - name: chat
     route_type: llm/v1/chat
```
Comment (harupy) [inline examples/gateway/openai/config.yaml:3]:
```suggestion
    endpoint_type: llm/v1/chat
```

should we use `endpoint`?

File: examples/gateway/openai/config.yaml [hunk 5]
```diff
@@ -1,4 +1,4 @@
-routes:
```
Comment (harupy) [inline examples/gateway/openai/config.yaml:1]: Can we show a deprecation warning if we find the following fields?

- `routes`
- `route_type`

Comment (harupy) [reply examples/gateway/openai/config.yaml:1]: I think we can insert a warning in `_load_route_config`.

File: mlflow/gateway/utils.py (hunk 3 above)
Comment (harupy) [reply mlflow/gateway/utils.py:33]: or:

```suggestion
    routes = config.get("routes") or config.get("endpoints") or []
```

File: mlflow/deployments/cli.py (hunk 2 above)
Comment (harupy) [reply mlflow/deployments/cli.py:493]: Filed #10470

File: examples/gateway/openai/config.yaml (hunk 4 above)
Comment (dbczumar) [reply examples/gateway/openai/config.yaml:3]: Reverting changes to this file

File: examples/gateway/openai/config.yaml [hunk 6]
```diff
@@ -1,4 +1,4 @@
...
+endpoints:
```
Comment (dbczumar) [inline examples/gateway/openai/config.yaml:1]:
```suggestion
routes:
```

File: mlflow/deployments/cli.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/deployments/cli.py:493]: Thanks! <3

File: mlflow/gateway/utils.py (hunk 3 above)
Comment (dbczumar) [reply mlflow/gateway/utils.py]: Committed - thanks!

**Review by harupy — APPROVED** [review]
LGTM!

File: examples/gateway/openai/config.yaml (hunk 5 above)
Comment (dbczumar) [reply examples/gateway/openai/config.yaml]: Added the warning and test :)

**Pull request merged**